power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/math_solver_dispatch.hpp

```diff
@@ -0,0 +1,156 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+// runtime dispatch for math solver
+// this can separate math solver into a different translation unit
+
+#pragma once
+
+#include "../calculation_parameters.hpp"
+#include "../common/common.hpp"
+#include "../common/exception.hpp"
+#include "../common/three_phase_tensor.hpp"
+#include "../common/timer.hpp"
+
+#include <memory>
+
+namespace power_grid_model {
+
+namespace math_solver {
+
+// forward declare YBus
+template <symmetry_tag sym> class YBus;
+
+template <template <symmetry_tag> class MathSolverType> struct math_solver_tag {};
+
+struct MathSolverDispatcher {
+    template <symmetry_tag sym> struct MathSolverDispatcherConfig {
```

Comment:
optionally, you can rename this to just `Config` since it's a member struct anyways